### PR TITLE
refactor: add m365 rag ts template

### PR DESCRIPTION
**Pull request by xiaolang124**
Add m365 rag ts template.

TODO: Update readme wiki part. Add question modal part.

image

### CODE REVIEW

File: templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                    // https://learn.microsoft.com/sharepoint/technical-reference/default-crawled-file-name-extensions-and-parsed-file-types
+                    queryString: `${graphQuery}`,
+                },
+                // This parameter is required only when searching with application permissions
+                // https://learn.microsoft.com/graph/search-concept-searchall
+                // region: "US",
+                },
+            ],
+        });
+        for (const value of response?.value ?? []) {
```
Comment (frankqianms) [inline templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl:74]: I wonder if response is something like 404, will it be null? I mean whether there will be a runtime error if you don't catch a possible error here?

File: templates/ts/custom-copilot-rag-microsoft365/teamsapp.local.yml.tpl [hunk 2]
```diff
@@ -0,0 +1,113 @@
+# yaml-language-server: $schema=https://aka.ms/teams-toolkit/1.0.0/yaml.schema.json
+# Visit https://aka.ms/teamsfx-v5.0-guide for details on this file
+# Visit https://aka.ms/teamsfx-actions for details on actions
+version: 1.0.0
```
Comment (frankqianms) [inline templates/ts/custom-copilot-rag-microsoft365/teamsapp.local.yml.tpl:4]: version needs to be upgraded to V1.3

File: templates/ts/custom-copilot-rag-microsoft365/teamsapp.yml.tpl [hunk 3]
```diff
@@ -0,0 +1,163 @@
+# yaml-language-server: $schema=https://aka.ms/teams-toolkit/1.0.0/yaml.schema.json
+# Visit https://aka.ms/teamsfx-v5.0-guide for details on this file
+# Visit https://aka.ms/teamsfx-actions for details on actions
+version: 1.0.0
```
Comment (frankqianms) [inline templates/ts/custom-copilot-rag-microsoft365/teamsapp.yml.tpl:4]: same here

File: templates/ts/custom-copilot-rag-microsoft365/package.json.tpl [hunk 4]
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "{{SafeProjectNameLowerCase}}",
+    "version": "1.0.0",
+    "msteams": {
+        "teamsAppId": null
+    },
+    "description": "Microsoft Teams Toolkit AI Chat Bot Sample with Teams AI Library",
```
Comment (kuojianlu) [inline templates/ts/custom-copilot-rag-microsoft365/package.json.tpl:7]: Update

File: templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    // Download the file from SharePoint
+    // https://docs.microsoft.com/en-us/graph/api/driveitem-get-content
+    private async downloadSharepointFile(
+        contentUrl: string
+    ): Promise<string | undefined> {
+        const encodedUrl = this.encodeSharepointContentUrl(contentUrl);
+        const fileContentResponse = await this.graphClient
+        .api(`/shares/${encodedUrl}/driveItem/content`)
```
Comment (kuojianlu) [inline templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl:120]: Missing indent

File: templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        .get();
+
+        return fileContentResponse;
+    }
+
+    private encodeSharepointContentUrl(webUrl: string): string {
+        const byteData = Buffer.from(webUrl, "utf-8");
+        const base64String = byteData.toString("base64");
+        return (
+        "u!" + base64String.replace("=", "").replace("/", "_").replace("+", "_")
```
Comment (kuojianlu) [inline templates/ts/custom-copilot-rag-microsoft365/src/app/graphDataSource.ts.tpl:131]: Indent

File: templates/ts/custom-copilot-rag-microsoft365/package.json.tpl [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        "@microsoft/microsoft-graph-client": "^3.0.1",
+        "@azure/search-documents": "^12.0.0",
+        "@microsoft/teams-ai": "^1.1.0",
+        "botbuilder": "^4.20.0",
+        "openai": "~4.28.4",
+        "restify": "^10.0.0"
+    },
+    "devDependencies": {
+        "@types/restify": "^8.5.5",
+        "@types/node": "^14.0.0",
```
Comment (kuojianlu) [inline templates/ts/custom-copilot-rag-microsoft365/package.json.tpl:38]: Apply this change: https://github.com/OfficeDev/TeamsFx/pull/11265

File: packages/fx-core/tests/question/create.test.ts [hunk 8]
```diff
@@ -1259,7 +1259,7 @@ describe("scaffold question", () => {
           } else if (question.name === QuestionNames.CustomCopilotRag) {
             const select = question as SingleSelectQuestion;
             const options = await select.dynamicOptions!(inputs);
-            assert.isTrue(options.length === 2);
+            assert.isTrue(options.length === 3);
             return ok({ type: "success", result: CustomCopilotRagOptions.customize().id });
```
Comment (frankqianms) [inline packages/fx-core/tests/question/create.test.ts:1263]: minor: would you mind help change `CustomCopilotRagOptions.customize().id` to `CustomCopilotRagOptions.azureAISearch().id` according to the test case name.

Comment (frankqianms) [reply packages/fx-core/tests/question/create.test.ts:1263]: maybe we could also raise a pr for all these test cases refractoring later.

**Review by kuojianlu — APPROVED** [review]

**Review by jayzhang — APPROVED** [review]
LGTM

**Pull request merged**